storage/scylla_api.h

```diff
@@ -0,0 +1,188 @@
+
+#define ADDRESS_1                                                               \
+  "ADDRESS100000LIKCEJTTIQOTTAWSQCCQQ9A9VOKIWRBWVPXMCGUENWVVMQAMPEIVHEQ9JXLCNZ" \
+  "OORVZT"
+#define ADDRESS_2                                                               \
+  "ADDRESS200000LIKCEJTTIQOTTAWSQCCQQ9A9VOKIWRBWVPXMCGUENWVVMQAMPEIVHEQ9JXLCNZ" \
+  "OORVZT"
+#define ADDRESS_3                                                               \
+  "ADDRESS300000LIKCEJTTIQOTTAWSQCCQQ9A9VOKIWRBWVPXMCGUENWVVMQAMPEIVHEQ9JXLCNZ" \
+  "OORVZT"
+#define ADDRESS_4                                                               \
+  "ADDRESS400000LIKCEJTTIQOTTAWSQCCQQ9A9VOKIWRBWVPXMCGUENWVVMQAMPEIVHEQ9JXLCNZ" \
+  "OORVZT"
+#define ADDRESS_5                                                               \
+  "ADDRESS500000LIKCEJTTIQOTTAWSQCCQQ9A9VOKIWRBWVPXMCGUENWVVMQAMPEIVHEQ9JXLCNZ" \
+  "OORVZT"
+#define ADDRESS_6                                                               \
+  "ADDRESS600000LIKCEJTTIQOTTAWSQCCQQ9A9VOKIWRBWVPXMCGUENWVVMQAMPEIVHEQ9JXLCNZ" \
+  "OORVZT"
+
+#define TRANSACTION_1                                                               \
+  "TRANSACTION100000LIKCEJTTIQOTTAWSQCCQQ9A9VOKIWRBWVPXMCGUENWVVMQAMPEIVHEQ9JXLCNZ" \
+  "OO"
+#define TRANSACTION_2                                                               \
+  "TRANSACTION200000LIKCEJTTIQOTTAWSQCCQQ9A9VOKIWRBWVPXMCGUENWVVMQAMPEIVHEQ9JXLCNZ" \
+  "OO"
+#define TRANSACTION_3                                                               \
+  "TRANSACTION300000LIKCEJTTIQOTTAWSQCCQQ9A9VOKIWRBWVPXMCGUENWVVMQAMPEIVHEQ9JXLCNZ" \
+  "OO"
+#define TRANSACTION_4                                                               \
+  "TRANSACTION400000LIKCEJTTIQOTTAWSQCCQQ9A9VOKIWRBWVPXMCGUENWVVMQAMPEIVHEQ9JXLCNZ" \
+  "OO"
+#define TRANSACTION_5                                                               \
+  "TRANSACTION500000LIKCEJTTIQOTTAWSQCCQQ9A9VOKIWRBWVPXMCGUENWVVMQAMPEIVHEQ9JXLCNZ" \
+  "OO"
+#define TRANSACTION_6                                                               \
+  "TRANSACTION600000LIKCEJTTIQOTTAWSQCCQQ9A9VOKIWRBWVPXMCGUENWVVMQAMPEIVHEQ9JXLCNZ" \
+  "OO"
+#define TRANSACTION_7                                                               \
+  "TRANSACTION700000LIKCEJTTIQOTTAWSQCCQQ9A9VOKIWRBWVPXMCGUENWVVMQAMPEIVHEQ9JXLCNZ" \
+  "OO"
+#define TRANSACTION_8                                                               \
+  "TRANSACTION800000LIKCEJTTIQOTTAWSQCCQQ9A9VOKIWRBWVPXMCGUENWVVMQAMPEIVHEQ9JXLCNZ" \
+  "OO"
+#define TRANSACTION_9                                                               \
+  "TRANSACTION900000LIKCEJTTIQOTTAWSQCCQQ9A9VOKIWRBWVPXMCGUENWVVMQAMPEIVHEQ9JXLCNZ" \
+  "OO"
+
+#define TIMESTAMP_1 1
+#define TIMESTAMP_2 2
+#define TIMESTAMP_3 3
+
+#define MESSAGE_1 "MESSAGE1"
+#define MESSAGE_2 "MESSAGE2"
+#define MESSAGE_3 "MESSAGE3"
+#define MESSAGE_4 "MESSAGE4"
+#define MESSAGE_5 "MESSAGE5"
+#define MESSAGE_6 "MESSAGE6"
+
+typedef struct Transaction_ {
```

Comment:
Is it really necessary to define `struct Transaction` here? Can you apply the technique of forward declaration?

Reply:
I think many get/set function for each struct member is needed since the caller won't know the implementation of the struct if applying the technique o forward declaration.
Is it fine or there's another better way?
For example.
set hash
```
status_t set_transaction_hash(Transaction* obj, cass_byte_t* hash, size_t length){
  if(length != NUM_FLEX_TRITS_HASH){
    return SC_STORAGE_INVAILD_INPUT;
  }
  memcpy(obj->transaction, hash, NUM_FLEX_TRITS_HASH);
  return SC_OK
}

cass_byte_t* ret_transaction_hash(Transaction* obj){
  return obj->transaction;
}
```
set address
```
status_t set_transaction_address(Transaction* obj, cass_byte_t* hash, size_t length){
  if(length != NUM_FLEX_TRITS_ADDRESS){
    return SC_STORAGE_INVAILD_INPUT;
  }
  memcpy(obj->address, hash, NUM_FLEX_TRITS_ADDRESS);
  return SC_OK
}
cass_byte_t* ret_transaction_hash(Transaction* obj){
  return obj->address;
}
```

And I would like to modify struct members to a fixed size as below. 
```
struct Transaction {
  cass_byte_t bundle[NUM_FLEX_TRITS_BUNDLE];
  cass_byte_t address[NUM_FLEX_TRITS_ADDRESS];
  cass_byte_t transaction[NUM_FLEX_TRITS_HASH];
  char message[NUM_FLEX_TRITS_MESSAGE];
  cass_int64_t value;
  cass_int64_t timestamp;
  cass_byte_t trunk[NUM_FLEX_TRITS_TRUNK];
  cass_byte_t branch[NUM_FLEX_TRITS_BRANCH];
};
```

Reply:
And actually I've wondered one question for a long time.
Is it bad if I use Marco to declare functions?
If I don't want to copy and paste then modify every function and check if there are typos.

For example.
```
#define DECLARE_SET_FUNCTION(VAR) \
  status_t set_transaction_##VAR(Transaction* obj, cass_byte_t* hash, size_t length){ \
    if(obj == NULL || hash == NULL || length != NUM_FLEX_TRITS_##VAR){ \
      return SC_STORAGE_INVAILD_INPUT; \
    } \
    memcpy(obj->##VAR, hash, NUM_FLEX_TRITS_##VAR); \
    return SC_OK \
  }
```
then

```
DECLARE_SET_FUNCTION(hash)
DECLARE_SET_FUNCTION(bundle)
DECLARE_SET_FUNCTION(address)
DECLARE_SET_FUNCTION(trunk)
DECLARE_SET_FUNCTION(branch)

```

Reply:
Using macros is straightforward, but you have to take care of type checking.